syntax = "proto3";

package qubit_engine;

option go_package = "github.com/perclft/QubitEngine/cli/internal/generated";

service QuantumCompute {
  rpc RunCircuit (CircuitRequest) returns (StateResponse) {}
  rpc StreamGates (stream GateOperation) returns (stream Measurement) {}
}

message CircuitRequest {
  int32 num_qubits = 1;
  repeated GateOperation operations = 2;
}

message GateOperation {
  enum GateType {
    HADAMARD = 0;
    PAULI_X = 1;
    CNOT = 2;
    MEASURE = 3; // New Operation
  }
  GateType type = 1;
  uint32 target_qubit = 2;
  uint32 control_qubit = 3;
  // Optional: Register to store the classical result (useful for complex circuits)
  uint32 classical_register = 4; 
}

message StateResponse {
  message ComplexNumber {
    double real = 1;
    double imag = 2;
  }
  repeated ComplexNumber state_vector = 1;
  
  // New: Return measured classical bits (e.g., Qubit 0 -> 1)
  map<uint32, bool> classical_results = 2; 
}

message Measurement {
  uint32 qubit_index = 1;
  bool result = 2;
  double probability = 3;
}
